Form/FormFlow.php

```diff
@@ -609,6 +645,16 @@ protected function bindFlow() {
 			$reset = true;
 		}
 
+		if ($this->getRequest()->isMethod('POST') && ! $this->dataManager->hasInstance($this)) {
+			// Form is stale, kill posted data and reset
+			$this->getRequest()->request->replace();
+			$reset = true;
+			$this->setStale(true);
+
+                        // Regenerate instance ID
+			$this->setInstanceId($this->determineInstanceId());
```

Comment:
It only regenerates the instance ID on a form expiration error, so it doesn't affect "redirect after submit" (and it still passes those tests). The purpose is so it will continue to give the error no matter how many times they resubmit. If you didn't regenerate the ID, then you could submit new information to the form, then back up again to an old, resubmit and confuse things. So just to clarify,

1) Page 1 -> Page 2 -> Page 3 -> Page 4 -> Submit
2) Back up to Page 4, resubmit. Gives form expiration error and resets the form.
3) Start second submit: Page 1 -> Page 2
4) Back up to Page 3 from earlier instance and submit. If we used the same instance ID, it's confused (or even worse, back up to first Page 2). But with the instance ID regenerated, it gives the proper expired form error.

That's the tricky case, but the simple case is actually just backing up to a page and resubmitting twice. Since for the initial page load it now initializes the data store with the instance ID, if you resubmit twice, the second time doesn't give the proper error.